datafusion/core/tests/sql/joins.rs

```diff
@@ -2040,8 +2040,8 @@ async fn left_semi_join() -> Result<()> {
         let physical_plan = state.create_physical_plan(&logical_plan).await?;
         let expected = if repartition_joins {
             vec![
-                "SortExec: [t1_id@0 ASC NULLS LAST]",
-                "  CoalescePartitionsExec",
+                "SortPreservingMergeExec: [t1_id@0 ASC NULLS LAST]",
+                "  SortExec: [t1_id@0 ASC NULLS LAST]",
```

Comment:
This is a cool plan -- it is the first time I know of that DataFusion crates plans using `SortPreservingMerge` directly 👍 

The rationale for this, as I understand it, is to support sorting in parallel. Internally the Sort operator does use a Merge if it has multiple input partitions. I wonder if we have any opinions about making the parallelism explicit in the plan (like this) or implict (within the operator)?

I think explicit in the plan is consistent with other parts of DataFusion (e.g. `CoalscePartionExec`)

fyi @tustvold

Reply:
In a previous PR for TopK-queries I added the check on a `LIMIT` to be present on purpose, 
The reason is that `SortPreservingMergeExec` + parallel `SortExec` is slower than `CoalescePartition` + `SortExec`   in my tests, so actually making queries run quite a bit slower. I think we have to use other means to parallelize sorts, such as range partitioning.

Reply:
Why `SortPreservingMergeExec` + parallel `SortExec` is slower than `CoalescePartitionsExec` + `SortExec`?  Maybe it depends on the input data volumes need to be sorted.
Do you have some test data and SQL ?

Reply:
I think the core idea of divide and conquer for parallel sorting and merging is very general. Also curious why the performance is worse than the single node global sort.

Reply:
what's more, the reason of depending on the limit to decide whether to change to parallel sort is also not strong. Sometimes, the limit number is very large and nothing can be pruned.

Reply:
Both excellent points.

* Limit looked like a relatively good heuristic, as one often provides a smaller value than the size of the input like 10, 100, 1000 or 10K but not 10M, 100M etc. One could fine tune the heuristic by only doing it on a "sufficiently" small limit. The limit causes the number of rows as input to `SortPreservingMergeExec` to be only `partitions * limit` (beside also speeding up the parallel sort).

* I am not sure whether we should expect `SortPreservingMergeExec` + parallel `SortExec` to be (much) faster than  `CoalescePartitionsExec` + `SortExec`.  `SortPreservingMergeExec` itself is running on one partition/thread, is a relative heavy operation and also needs to wait on the sorted input partitions to be all ready.

Maybe with some fine-tuning (or maybe already after https://github.com/apache/arrow-datafusion/pull/4301) it is possible it is faster than a single-threaded `SortExec` in certain cases, but other approaches like using range partitioning will be better for query parallelism.